### ISSUE TITLE
ApplicativePlus some/many are nonsense

### ISSUE THREAD

**Issue by djspiewak** [scalaz/scalaz#1097]
First off, they don't work and never have.  To wit:

``` scala
import scalaz._
import Scalaz._

ApplicativePlus[Option].some(None)       // EAGERLY explodes with stack overflow
```

The trampolining in the implementation is misguided and only compiles because Scala doesn't bound infinite recursion (if you look closely, you'll realize there is no introduction of the trampoline value).  More importantly, just tossing trampoline on the _idea_ of these functions doesn't make any sense, since you would really need to thread the trampoline through `apply2` in order to get any stack safety.  However, that is clearly nonsense since we need to use the `apply2` from `Applicative[F]`, and so trampolining is already a non-starter.

Even more fundamentally though, these functions make exactly zero sense in Scala.  Think about the signature:

``` scala
F[A] => F[List[A]]
```

There are only two rational implementations of the `many` function (with infinite variants on the second implementation):
- `many(fa: F[A]): F[List[A]] = fa map { _ => Nil }`
- `many(fa: F[A]): F[List[A]] = fa map { _ :: Nil }`

You can repeat as many _finite_ times as you want in the second one, but it's still a broadly arbitrary wrapping.  In Haskell, presumably, these functions generate infinite lists (that certainly looks like what the implementation is _trying_ to do).  In Scala, `List` is finite.  There is nothing to bound the recursion.

You could reimplement these functions with `Stream`, though they would only be stack safe for effects `F` which are themselves stack safe (e.g. `Free` good, `Option` bad).  It would be better to simply remove them though, since a) no one has even LOOKED at using them since mid-2012 (which we know since they have been broken in all cases since that date), and c) the functionality they provide is arguably not useful in a point-heavy language with eager semantics.

**Comment by puffnfresh** [comment]
`some` and `many` are also non-termination for a lot of data structures in Haskell (e.g. `[]` and `Maybe`) so I think scalaz' behaviour is consistent and irrelevant of Scala being eager by-default.

But a version of `Alternative` adds a law which ensures `some` and `many` do terminate would be useful!

**Comment by stew** [comment]
This can't possibly work if you never ever call any of the Trampoline constructors, right?

**Comment by bmjames** [comment]
I started a [discussion about this on IRC](http://ircbrowse.net/browse/scalaz?id=53606&timestamp=1417085598#t1417085598) a while ago, and proposed this implementation: https://github.com/bmjames/scalaz/commit/c0cbb6c20479508c06ba93b6bc42438b4e2b9ce8

It looks almost too naive to work, but apparently (judging from the tests I wrote in that patch) it does terminate and do something useful for an `ApplicativePlus[F]` as long as `F` is trampolined.

**Comment by bmjames** [comment]
I'll also leave this idea here:

I _think_ (I haven't written it yet) that if `F` is also a `Monad`, it would be possible to provide an implementation that lifts `F` into `Free`, thus discharging the need to trampoline your own `F`. So, if it is decided that there is no sufficiently useful implementation of `some` and `many` in Scala for `ApplicativePlus`, there may still be one for `MonadPlus`.

**Comment by djspiewak** [comment]
@bmjames I thought about that implementation.  It requires a particularly generous interpretation of `empty` (namely, `empty` would not be a unique value, which weakens the laws considerably).  Such an implementation would probably be useful most useful with an effect like `Task`.

@puffnfresh Considering that these functions *immediately* fail with any input (you can toss `null` in there and it fails just the same), I'm fairly certain that the behavior is not consistent with anything other than insanity.  Removing the trampoline usage entirely *would* produce an implementation which is broadly consistent with Haskell, modulo stack safety and eagerness issues and effects with a unique `plus` identity.

**Comment by fommil** [comment]
:laughing: YES, thank you @djspiewak I have spent the last 20 minutes working out what the hell these things are for. I can't even figure out what they are supposed to do differently to each other!

I'm going to omit them from my book.

**Comment by fommil** [comment]
I figured out why these are bust and will fix as part of #1866

**Comment by fommil** [comment]
TL;DR delete them, they belong on Alternative, which should be it's own typeclass not an alias

**Comment by tonymorris** [comment]
Make sure `some` uses `NonEmpty`. Yes these can work in Scala.

**Comment by fommil** [comment]
I made Applicative have lazy semantics recently, so that should help.
